intents/intent.go

```diff
@@ -66,48 +78,77 @@ func (intent *Intent) Hash() ([]byte, error) {
 	return crypto.Keccak256(packetBytes), nil
 }
 
-func (intent *Intent) Signers() []string {
+func (intent *Intent) Signers(verifierGetter IntentVerifierGetter) []string {
 	var signers []string
 
 	for _, signature := range intent.signatures {
-		if intent.isValidSignature(signature.Session, signature.Signature) {
-			signers = append(signers, signature.Session)
+		sessionVerifier, err := verifierGetter(signature.SessionId)
+		if err != nil {
+			continue
+		}
+
+		if intent.isValidSignature(sessionVerifier, signature.Signature) {
+			signers = append(signers, sessionVerifier)
 		}
 	}
 
 	return signers
 }
 
-func (intent *Intent) IsValid() bool {
+func (intent *Intent) IsValid(verifierGetter IntentVerifierGetter) (bool, error) {
 	// Check if the packet is valid
 	var packet packets.BasePacket
 	err := json.Unmarshal(intent.Packet, &packet)
 	if err != nil {
-		return false
+		return false, fmt.Errorf("intent: %w", ErrInvalidPacket)
 	}
 
 	// OpenSession packets do not require signatures
 	if packet.Code == packets.OpenSessionPacketCode {
-		return packet.IsValid()
+		if ok, err := packet.IsValid(); !ok {
+			return false, fmt.Errorf("intent: %w", err)
+		}
+		return true, nil
 	}
 
 	// Check if there are any signatures
 	if len(intent.signatures) == 0 {
-		return false
+		return false, fmt.Errorf("intent: %w", ErrNoSignatures)
 	}
 
 	// Check if all signatures are valid
 	for _, signature := range intent.signatures {
-		if !intent.isValidSignature(signature.Session, signature.Signature) {
-			return false
+		sessionVerifier, err := verifierGetter(signature.SessionId)
+		if err != nil {
+			return false, fmt.Errorf("intent: %w", err)
+		}
+
+		if !intent.isValidSignature(sessionVerifier, signature.Signature) {
+			return false, fmt.Errorf("intent: %w", ErrInvalidSignature)
 		}
 	}
 
 	// Check if the packet is valid
-	return packet.IsValid()
+	if ok, err := packet.IsValid(); !ok {
+		return false, fmt.Errorf("intent: %w", err)
+	}
+	return true, nil
+}
+
+func (intent *Intent) isValidSECP256R1Session(session string, signature string) bool {
+	return strings.HasPrefix(session, "r1:") && strings.HasPrefix(signature, "r1:")
 }
 
 func (intent *Intent) isValidSignature(session string, signature string) bool {
+	if intent.isValidSECP256R1Session(session, signature) {
+		return intent.isValidSignatureSECP256R1(session, signature)
+	} else {
+		return intent.isValidSignatureSPECP256K1(session, signature)
+	}
+}
+
+// isValidSignatureSPECP256K1 checks if the signature is valid for the given secp256k1 session
+func (intent *Intent) isValidSignatureSPECP256K1(session string, signature string) bool {
```

Comment:
typo here.. you wrote `SPECP256K1` but should be `SECP256K1` .. however, as a short-hand, I think its fine to just use the name.. `isValidSignatureP256K1` and `isValidSignatureP256R1`